pkg/xpkg/build.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package xpkg
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/empty"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/pkg/errors"
+
+	"github.com/crossplane/crossplane-runtime/pkg/parser"
+)
+
+const (
+	errParserPackage = "failed to parse package"
+	errLintPackage   = "failed to lint package"
+	errInitBackend   = "failed to initialize package parsing backend"
+	errCopyStream    = "failed to copy stream into buffer"
+	errTarFromStream = "failed to build tarball from package stream"
+	errLayerFromTar  = "failed to convert tarball to image layer"
+)
+
+// Build compiles a Crossplane package from an on-disk package.
+func Build(ctx context.Context, b parser.Backend, p parser.Parser, l parser.Linter) (v1.Image, error) {
+	// Get YAML stream.
+	r, err := b.Init(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, errInitBackend)
+	}
+	defer func() { _ = r.Close() }()
+
+	// Copy stream once to parse and once write to tarball.
+	buf := new(bytes.Buffer)
+	pkg, err := p.Parse(ctx, ioutil.NopCloser(io.TeeReader(r, buf)))
+	if err != nil {
+		return nil, errors.Wrap(err, errParserPackage)
+	}
+	if err := l.Lint(pkg); err != nil {
+		return nil, errors.Wrap(err, errLintPackage)
+	}
+	_, err = io.Copy(buf, r)
+	if err != nil {
+		return nil, errors.Wrap(err, errCopyStream)
+	}
```

Comment:
Nit: This (and the instance below) could be inlined to `if _, err := ...; err != nil`. (Feels like there should be a linter for this.)